### ISSUE TITLE
Ignore nested components in no-multi-comp

### ISSUE THREAD

**Issue by remcohaszing** [yannickcr/eslint-plugin-react#2842]
I don’t think `react/no-multi-comp` should trigger on nested components or should have an option to disable it. A practical example:

```jsx
import { FormattedMessage } from 'react-intl';

export function App() {
  return (
    <FormattedMessage
      values={{
        link:
          // The next line triggers react/no-multi-comp
          (text: string) => <a>{text}</a>,
      }}
    />
  );
}
```

I suggest to add an option `ignoreNested` which ignores any components that aren’t declared as part of the `Program` node body.

**Comment by ljharb** [comment]
The issue there isn’t that it’s nested - it’s that it’s treated as a component despite being used in an object literal with a key name that can’t be a custom component name.

Our component detection should be improved here.

**Comment by remcohaszing** [comment]
> The issue there isn’t that it’s nested - it’s that it’s treated as a component despite being used in an object literal with a key name that can’t be a custom component name.

I like that explanation even better.
